File: cpp_backup/src/storage/wal.cpp
#include "flux/storage/wal.hpp"
#include "flux/util/filesystem.hpp"
#include <fstream>
#include <fmt/format.h>

namespace flux {

WAL::WAL(Path wal_dir) : wal_dir_(std::move(wal_dir)) {
}

Result<uint64_t> WAL::begin_transaction() {
    uint64_t txn_id = next_txn_id_++;
    
    // Create transaction directory
    Path txn_path = get_transaction_path(txn_id);
    auto result = Filesystem::create_directories(txn_path);
    if (!result) {
        return flux::unexpected(result.error());
    }
    
    // Write transaction header
    std::string header = fmt::format("txn:{}\nstatus:active\n", txn_id);
    Bytes data(header.begin(), header.end());
    
    auto write_result = Filesystem::write_file(txn_path / "header", data);
    if (!write_result) {
        return flux::unexpected(write_result.error());
    }
    
    return txn_id;
}

Result<void> WAL::log_operation(uint64_t txn_id, const std::string& operation, const Bytes& data) {
    Path txn_path = get_transaction_path(txn_id);
    
    if (!Filesystem::exists(txn_path)) {
        return flux::unexpected(fmt::format("Transaction {} not found", txn_id));
    }
    
    // Append operation to log
    Path log_path = txn_path / "operations";
    
    // Format: <operation_len><operation><data_len><data>
    Bytes log_entry;
    
    uint32_t op_len = operation.size();
    log_entry.insert(log_entry.end(),
        reinterpret_cast<uint8_t*>(&op_len),
        reinterpret_cast<uint8_t*>(&op_len) + sizeof(op_len));
    log_entry.insert(log_entry.end(), operation.begin(), operation.end());
    
    uint64_t data_len = data.size();
    log_entry.insert(log_entry.end(),
        reinterpret_cast<uint8_t*>(&data_len),
        reinterpret_cast<uint8_t*>(&data_len) + sizeof(data_len));
    log_entry.insert(log_entry.end(), data.begin(), data.end());
    
    // Append to file (not atomic, but that's okay for WAL)
    std::ofstream file(log_path, std::ios::binary | std::ios::app);
    if (!file) {
        return flux::unexpected(fmt::format("Failed to open WAL log: {}", log_path.string()));
    }
    
    file.write(reinterpret_cast<const char*>(log_entry.data()), log_entry.size());
    
    return {};
}

Result<void> WAL::commit_transaction(uint64_t txn_id) {
    Path txn_path = get_transaction_path(txn_id);
    
    if (!Filesystem::exists(txn_path)) {
        return flux::unexpected(fmt::format("Transaction {} not found", txn_id));
    }
    
    // Mark as committed
    std::string header = fmt::format("txn:{}\nstatus:committed\n", txn_id);
    Bytes data(header.begin(), header.end());
    
    auto result = Filesystem::write_file(txn_path / "header", data);
    if (!result) {
        return result;
    }
    
    // Clean up transaction directory
    return Filesystem::remove(txn_path);
}

Result<void> WAL::rollback_transaction(uint64_t txn_id) {
    Path txn_path = get_transaction_path(txn_id);
    
    if (!Filesystem::exists(txn_path)) {
        return {};  // Already rolled back
    }
    
    // Simply remove transaction directory
    return Filesystem::remove(txn_path);
}

Result<void> WAL::recover() {
    // Check for incomplete transactions
    if (!Filesystem::exists(wal_dir_)) {
        return {};
    }
    
    auto entries_result = Filesystem::list_directory(wal_dir_);
    if (!entries_result) {
        return flux::unexpected(entries_result.error());
    }
    
    // Roll back any active transactions
    for (const auto& entry : *entries_result) {
        if (Filesystem::is_directory(entry)) {
            // This is a transaction directory
            // For now, just remove it (rollback)
            // In a full implementation, we'd replay or rollback based on status
            auto result = Filesystem::remove(entry);
            if (!result) {
                return result;
            }
        }
    }
    
    return {};
}

Path WAL::get_transaction_path(uint64_t txn_id) const {
    return wal_dir_ / fmt::format("txn_{:016x}", txn_id);
}

} // namespace flux

File: src/storage/wal.zig
const std = @import("std");
const filesystem = @import("../util/filesystem.zig");

const Filesystem = filesystem.Filesystem;

/// WAL - Write-Ahead Log for transactional operations
pub const WAL = struct {
    wal_dir: []const u8,
    next_txn_id: u64,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, wal_dir: []const u8) WAL {
        return .{
            .allocator = allocator,
            .wal_dir = wal_dir,
            .next_txn_id = 1,
        };
    }

    /// Begin transaction
    pub fn beginTransaction(self: *WAL) !u64 {
        const txn_id = self.next_txn_id;
        self.next_txn_id += 1;

        const txn_path = try self.getTransactionPath(txn_id);
        defer self.allocator.free(txn_path);

        // Create empty transaction file
        try Filesystem.writeFile(txn_path, "");

        return txn_id;
    }

    /// Log operation
    pub fn logOperation(self: *WAL, txn_id: u64, operation: []const u8, data: []const u8) !void {
        const txn_path = try self.getTransactionPath(txn_id);
        defer self.allocator.free(txn_path);

        // Append operation to transaction log
        var log_data = std.ArrayList(u8).init(self.allocator);
        defer log_data.deinit();

        try log_data.writer().print("{s}\n", .{operation});
        try log_data.appendSlice(data);
        try log_data.append('\n');

        // Append to file
        const file = try std.fs.cwd().openFile(txn_path, .{ .mode = .write_only });
        defer file.close();

        try file.seekFromEnd(0);
        try file.writeAll(log_data.items);
    }

    /// Commit transaction
    pub fn commitTransaction(self: *WAL, txn_id: u64) !void {
        const txn_path = try self.getTransactionPath(txn_id);
        defer self.allocator.free(txn_path);

        // Remove transaction log (operation completed)
        try Filesystem.remove(txn_path);
    }

    /// Rollback transaction
    pub fn rollbackTransaction(self: *WAL, txn_id: u64) !void {
        const txn_path = try self.getTransactionPath(txn_id);
        defer self.allocator.free(txn_path);

        // Remove transaction log
        Filesystem.remove(txn_path) catch {};
    }

    /// Recover from crash (placeholder - full implementation would replay logs)
    pub fn recover(self: *WAL) !void {
        _ = self;
        // TODO: Implement recovery by replaying uncommitted transactions
    }

    fn getTransactionPath(self: *WAL, txn_id: u64) ![]u8 {
        const filename = try std.fmt.allocPrint(self.allocator, "txn_{d}.log", .{txn_id});
        defer self.allocator.free(filename);

        const parts = [_][]const u8{ self.wal_dir, filename };
        return try std.fs.path.join(self.allocator, &parts);
    }
};
